examples/animatediff/ad/models/diffusion/ddpm.py

```diff
@@ -351,85 +271,144 @@ def decode_first_stage(self, z):
+        # optional: for some conditions, concat to latents, or add to time embedding
+        if self.cond_stage_trainable:
+            text_emb = self.cond_stage_model(text_tokens)
         else:
-            return x_recon
+            text_emb = ops.stop_gradient(self.cond_stage_model(text_tokens))
+        cond = {"c_crossattn": text_emb}
+
+        return cond
+
+    def construct(self, x: ms.Tensor, text_tokens: ms.Tensor, control=None, **kwargs):
+        """
+        Video diffusion model forward and loss computation for training
+
+        Args:
+            x: pixel values of video frames, resized and normalized to shape [bs, F, 3, 256, 256]
+            text: text tokens padded to fixed shape [bs, 77]
+            control: other conditions for future extension
 
-    def get_input(self, x, c):
-        if len(x.shape) == 3:
-            x = x[..., None]
-        x = self.transpose(x, (0, 3, 1, 2))
-        z = ops.stop_gradient(self.get_first_stage_encoding(self.encode_first_stage(x)))
-        return z, c
+        Returns:
+            loss
 
-    def construct(self, x, c):
+        Notes:
+            - inputs should matches dataloder output order
+            - assume unet3d input/output shape: (b c f h w)
+                unet2d input/output shape: (b c h w)
+        """
+
+        # 1. get image/video latents z using vae
+        z = self.get_latents(x)
+
+        # 2. sample timestep and add noise to latents
         t = self.uniform_int(
             (x.shape[0],), Tensor(0, dtype=mstype.int32), Tensor(self.num_timesteps, dtype=mstype.int32)
         )
-        x, c = self.get_input(x, c)
-        c = self.get_learned_conditioning_fortrain(c)
-        return self.p_losses(x, c, t)
+        noise = ops.randn_like(z)
+        noisy_latents = self.add_noise(z, noise, t)
+
+        # 3. get condition embeddings
+        cond = self.get_condition_embeddings(text_tokens, control)
 
-    def p_losses(self, x_start, cond, t, noise=None, **kwargs):
-        noise = msnp.randn(x_start.shape)
-        x_noisy = self.q_sample(x_start=x_start, t=t, noise=noise)
+        # 4.  unet forward, predict conditioned on conditions
         model_output = self.apply_model(
-            x_noisy,
+            noisy_latents,
             t,
             cond=cond,
             **kwargs,
         )
 
-        if self.parameterization == "x0":
-            target = x_start
-        elif self.parameterization == "eps":
+        # 5. compute loss
+        if self.parameterization == "eps":
             target = noise
-        elif self.parameterization == "velocity":
-            # target = sqrt_alpha_cum * noise - sqrt_one_minus_alpha_prod * x_start
-            target = self.get_velocity(x_start, noise, t)  # TODO: parse train step from randint
         else:
             raise NotImplementedError()
 
-        loss_simple = self.get_loss(model_output, target, mean=False).mean([1, 2, 3])
+        loss = self.mse_mean(target, model_output)
 
+        """
```

Comment:
`self.compute_loss` and `self.reduce_loss` are commented here and thus never called, but their definitions are not deleted. Is it intended or not?

Reply:
not used. leave for method improvement